src/provider/provider_cuda.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+
+#include <umf.h>
+#include <umf/providers/provider_cuda.h>
+
+#include "cuda.h"
+
+#include "base_alloc_global.h"
+#include "utils_assert.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+#include "utils_load_library.h"
+#include "utils_log.h"
+#include "utils_sanitizers.h"
+
+typedef struct cu_memory_provider_t {
+    CUcontext context;
+    CUdevice device;
+    umf_usm_memory_type_t memory_type;
+    size_t min_alignment;
+} cu_memory_provider_t;
+
+typedef struct cu_ops_t {
+    CUresult (*cuMemGetAllocationGranularity)(
+        size_t *granularity, const CUmemAllocationProp *prop,
+        CUmemAllocationGranularity_flags option);
+    CUresult (*cuMemAlloc)(CUdeviceptr *dptr, size_t bytesize);
+    CUresult (*cuMemAllocHost)(void **pp, size_t bytesize);
+    CUresult (*cuMemAllocManaged)(CUdeviceptr *dptr, size_t bytesize,
+                                  unsigned int flags);
+    CUresult (*cuMemFree)(CUdeviceptr dptr);
+    CUresult (*cuMemFreeHost)(void *p);
+
+    CUresult (*cuGetErrorName)(CUresult error, const char **pStr);
+    CUresult (*cuGetErrorString)(CUresult error, const char **pStr);
+} cu_ops_t;
+
+static cu_ops_t g_cu_ops;
+static UTIL_ONCE_FLAG cu_is_initialized = UTIL_ONCE_FLAG_INIT;
+static bool Init_cu_global_state_failed;
+
+#define TLS_MSG_BUF_LEN 1024
+
+typedef struct cu_last_native_error_t {
+    CUresult native_error;
+    char msg_buff[TLS_MSG_BUF_LEN];
+} cu_last_native_error_t;
+
+static __TLS cu_last_native_error_t TLS_last_native_error;
+
+static void cu_store_last_native_error(CUresult native_error) {
+    TLS_last_native_error.native_error = native_error;
```

Comment:
Please read how it is used in other providers and implement it corectly (we need provider specyfic error codes, error message string etc....)

Reply:
I store the error code here but provide to the user the string with the error message in cu_memory_provider_get_last_native_error()